mountpoint-s3/src/inode.rs

```diff
@@ -382,77 +370,93 @@ impl SuperblockInner {
+    pub fn update_from_remote(
         &self,
         parent_ino: InodeNo,
         name: &str,
-        stat: InodeStat,
-        kind: InodeKind,
-    ) -> Result<Inode, InodeError> {
-        let parent = {
-            let inodes = self.inodes.read().unwrap();
-            inodes
-                .get(&parent_ino)
-                .cloned()
-                .ok_or(InodeError::InodeDoesNotExist(parent_ino))?
-        };
+        remote: Option<RemoteLookup>,
+    ) -> Result<LookedUp, InodeError> {
+        let parent = self.get(parent_ino)?;
 
         // TODO what if a file was overwritten by a directory on the server side?
         if parent.kind() != InodeKind::Directory {
             return Err(InodeError::NotADirectory(parent_ino));
         }
 
-        // Fast path: try with only a read lock on the directory first
+        // Fast path: try with only a read lock on the directory first.
         {
             let parent_state = parent.inner.sync.read().unwrap();
-            if let Some(inode) = Self::update_if_present(parent_ino, &parent_state, name, kind, &stat)? {
-                return Ok(inode);
+            match Self::try_update_child(&parent_state, name, &remote)? {
+                UpdateStatus::Neither => return Err(InodeError::FileDoesNotExist),
+                UpdateStatus::Updated(lookedup) => return Ok(lookedup),
+                _ => {} // Fallback, we need a write lock to update the parent.
             }
         }
 
         // If the fast path failed, take the write lock. We first have to try the update again, as
         // a racing writer might have beat us to the lock after our fast path attempt.
         let mut parent_state = parent.inner.sync.write().unwrap();
-        if let Some(inode) = Self::update_if_present(parent_ino, &parent_state, name, kind, &stat)? {
-            return Ok(inode);
+        match Self::try_update_child(&parent_state, name, &remote)? {
+            UpdateStatus::Neither => Err(InodeError::FileDoesNotExist),
+            UpdateStatus::Updated(lookedup) => Ok(lookedup),
+            UpdateStatus::LocalOnly(inode) => {
+                match &mut parent_state.kind_data {
+                    InodeKindData::File {} => unreachable!("we know parent is a directory"),
+                    InodeKindData::Directory {
+                        children,
+                        writing_children,
+                    } => {
+                        if writing_children.contains(&inode.ino()) {
+                            // Return the local inode.
+                            let stat = inode.inner.sync.read().unwrap().stat.clone();
+                            return Ok(LookedUp { inode, stat });
+                        }
+                        // Remove from children.
+                        // TODO: also handle inode in [Self::inodes].
+                        children.remove(name);
+                    }
+                }
+                Err(InodeError::FileDoesNotExist)
```

Comment:
Maybe a little clearer with less fallthrough:

```suggestion
                match &mut parent_state.kind_data {
                    InodeKindData::File {} => unreachable!("we know parent is a directory"),
                    InodeKindData::Directory {
                        children,
                        writing_children,
                    } => {
                        if writing_children.contains(&inode.ino()) {
                            // Return the local inode.
                            let stat = inode.inner.sync.read().unwrap().stat.clone();
                            Ok(LookedUp { inode, stat })
                        } else {
                            // Remote no longer exists, so remove from local children
                            // TODO: also handle inode in [Self::inodes].
                            children.remove(name);
                            Err(InodeError::FileDoesNotExist)
                        }
                    }
                }
```